### PR TITLE
WebSocket use gorilla/websocket

### PR DESCRIPTION
**Pull request by juliens**
related to:

#17
#20
#21
#22
#26
#27
#28
#31

### CODE REVIEW

File: forward/fwd.go [hunk 1]
```diff
@@ -286,57 +291,58 @@ func (f *httpForwarder) serveWebSocket(w http.ResponseWriter, req *http.Request,
...
 
+	dialer := websocket.DefaultDialer
 	if outReq.URL.Scheme == "wss" && f.tlsClientConfig != nil {
-		f.log.Debugf("vulcand/oxy/forward/websocket: Dialing secure (tls) tcp connection to host %s with TLS Client Config %v", host, f.tlsClientConfig)
-		targetConn, err = tls.Dial("tcp", host, f.tlsClientConfig)
-	} else {
-		f.log.Debugf("vulcand/oxy/forward/websocket: Dialing insecure (non-tls) tcp connection to host %s", host)
-		targetConn, err = net.Dial("tcp", host)
+		dialer.TLSClientConfig = f.tlsClientConfig.Clone()
+		//WebSocket is only in http/1.1
```
Comment (mmatur) [inline forward/fwd.go]: Could you please add a space after `//`

File: forward/fwd.go [hunk 2]
```diff
@@ -286,57 +291,58 @@ func (f *httpForwarder) serveWebSocket(w http.ResponseWriter, req *http.Request,
...
+				log.Errorf("vulcand/oxy/forward/websocket: Failed to forward response")
+				ctx.errHandler.ServeHTTP(w, req, err)
+				return
+			}
+		}
 		return
 	}
-	underlyingConn, _, err := hijacker.Hijack()
+
+	//Only the targetConn choose to CheckOrigin or not
```
Comment (mmatur) [inline forward/fwd.go]: Could you please add a space after `//`

File: forward/fwd.go [hunk 3]
```diff
@@ -347,21 +353,27 @@ func (f *httpForwarder) serveWebSocket(w http.ResponseWriter, req *http.Request,
...
-	go replicate(underlyingConn, targetConn, "client", "backend")
-	err = <-errc // One goroutine complete
-	f.log.Infof("vulcand/oxy/forward/websocket: first proxying connection closed: %v", err)
-	err = <-errc // Both goroutines complete
-	f.log.Infof("vulcand/oxy/forward/websocket: second proxying connection closed: %v", err)
+
+	go replicate(targetConn.UnderlyingConn(), underlyingConn.UnderlyingConn(), "backend", "client")
+
+	// Try to read the first message
+	t, msg, err := targetConn.ReadMessage()
```
Comment (mmatur) [inline forward/fwd.go]: Could you please rename `t` to `msgType`

File: forward/fwd.go [hunk 4]
```diff
@@ -347,21 +353,27 @@ func (f *httpForwarder) serveWebSocket(w http.ResponseWriter, req *http.Request,
...
+		log.Errorf("vulcand/oxy/forward/websocket: Couldn't read first message : %v", err)
+	} else {
+		underlyingConn.WriteMessage(t, msg)
+	}
+
+	go replicate(underlyingConn.UnderlyingConn(), targetConn.UnderlyingConn(), "client", "backend")
+	<-errc
 }
 
 // copyRequest makes a copy of the specified request.
```
Comment (mmatur) [inline forward/fwd.go]: Could you please fix your comment `// copyWebSocketRequest makes a copy of the specified web socket request`

File: forward/fwd.go [hunk 5]
```diff
@@ -372,24 +384,26 @@ func (f *httpForwarder) copyWebSocketRequest(req *http.Request) (outReq *http.Re
...
+		}
+		outReq.URL.RawQuery = requestURI.RawQuery
 	}
 
-	// Overwrite close flag so we can keep persistent connection for the backend servers
-	outReq.Close = false
+	outReq.URL.Host = req.URL.Host
 
 	outReq.Header = make(http.Header)
+	//gorilla websocket use this header to set the request.Host tested in checkSameOrigin
```
Comment (mmatur) [inline forward/fwd.go]: Could you please add space after `//`

File: forward/fwd_websocket_test.go [hunk 6]
```diff
@@ -0,0 +1,500 @@
...
+
+	req, err := http.NewRequest(http.MethodGet, "ws://127.0.0.1/ws", nil)
+	c.Assert(err, IsNil)
+
+	req.Header.Add("upgrade", "websocket")
+	req.Header.Add("Connection", "upgrade")
+
+	req.Write(conn)
+
+	//First request works with 400
```
Comment (mmatur) [inline forward/fwd_websocket_test.go]: Could you please add a space after `//`

**Review by mmatur — CHANGES_REQUESTED** [review]

**Review by mmatur — APPROVED** [review]
LGTM :clap:

**Review by nmengin — APPROVED** [review]
LGTM

**Review by ldez — APPROVED** [review]
LGTM

**Pull request merged**